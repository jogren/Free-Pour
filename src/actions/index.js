export const showSelectDrink = targetDrink => ({
  type: 'SHOW_SELECT_DRINK',
  targetDrink
});

export const hideSelectedDrink = () => ({
  type: 'HIDE_SELECTED_DRINK'
});

export const setCurrentCocktails = cocktails => ({
  type: 'SET_CURRENT_COCKTAILS',
  cocktails
});