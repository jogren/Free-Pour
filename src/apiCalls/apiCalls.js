export const fetchPopularCocktails = async () => {
  const url = 'https://www.thecocktaildb.com/api/json/v2/8673533/popular.php';
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const popularCocktails = await response.json()
    return popularCocktails.drinks.map(drink => ({
      strDrink: drink.strDrink,
      strDrinkThumb: drink.strDrinkThumb,
    }))
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchCocktailsByGenre = async (type) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/filter.php?i=${type}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by genre')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch (error) {
    throw new Error(error.message)

  }
}

export const fetchCocktailsBySearch = async (text) => {
  const url = `https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${text}`
  try {
    const response = await fetch(url)
    if(!response.ok) {
      throw new Error('There was an issue fetching your data by name')
    }
    const cocktails = await response.json();
    return cocktails.drinks
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchAllIngredients = async () => {
  const url = '  https://www.thecocktaildb.com/api/json/v1/1/list.php?i=list';
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error('There was an issue fetching all ingredients')
    }
    const ingredients = await response.json();
    return ingredients.drinks.map(drink => drink.strIngredient1)
  } catch(error) {
    throw new Error(error.message)
  }
}

export const fetchMoreDrinkInfo = async (name) => {
  try {
    const response = await fetch(`https://www.thecocktaildb.com/api/json/v1/1/search.php?s=${name.split(' ').join('_')}`)
    if (!response.ok) {
      throw new Error('There was an error fetch your data')
    }
    const cocktail = await response.json();
    const cleanedCocktail = cocktail.drinks.map(drink => ({
      name: drink.strDrink,
      glass: drink.strGlass,
      instructions: drink.strInstructions,
      image: drink.strDrinkThumb,
      ingredients: [
        { measure: drink.strMeasure1, ingredient: drink.strIngredient1 },
        { measure: drink.strMeasure2, ingredient: drink.strIngredient2 },
        { measure: drink.strMeasure3, ingredient: drink.strIngredient3 },
        { measure: drink.strMeasure4, ingredient: drink.strIngredient4 },
        { measure: drink.strMeasure5, ingredient: drink.strIngredient5 },
        { measure: drink.strMeasure6, ingredient: drink.strIngredient6 }
      ]
    }))
    return cleanedCocktail[0]
  } catch(error) {
    throw new Error(error.message)
  }
}