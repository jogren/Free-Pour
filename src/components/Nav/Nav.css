nav {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: #FF6961;
  height: 50px;
  padding: 10px;
}