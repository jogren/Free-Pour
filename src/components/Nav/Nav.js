import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchCocktailsBySearch } from '../../apiCalls/apiCalls';
import { setCurrentCocktails } from '../../actions';
import { NavLink } from 'react-router-dom';
import './Nav.css';

export class Nav extends Component {
  constructor() {
    super();
    this.state = {
      search: '',
      searchHasErrored: '',
      favoritesHasErrored: ''
    }
  }

  handleSearch = e => {
    this.setState({ searchHasErrored: '' })
    this.setState({ [e.target.name]: e.target.value })
  }

  getCocktailsBySearch = async () => {
    try {
      const searchedCocktails = await fetchCocktailsBySearch(this.state.search)
      if (searchedCocktails === null) {
        this.setState({ searchHasErrored: 'Please try again!' })
      } else {
        this.props.setCurrentCocktails(searchedCocktails)
      }
      this.setState({ search: '' })
    } catch(error) {
      throw new Error(error.message)
    }
  }

  render() {
    const { getCocktailsByGenre, favoriteCocktails } = this.props
    return (
        <nav>
            <div className="Nav_div">
              <input 
                className="Nav_input-search"
                autoComplete="off"
                placeholder="Search for a drink..."
                name="search"
                value={this.state.search}
                onChange={(e) => this.handleSearch(e)}
                />
            <button onClick={this.getCocktailsBySearch} disabled={!this.state.search}>Submit</button>
              { this.state.searchHasErrored && <p>Please Check your Spelling!</p>}
            </div>
            <select onChange={(e) => getCocktailsByGenre(e.target.value)} className="select-container">
              <option value="">Select Genre:</option>
              <option value="vodka">Vodka</option>
              <option value="bourbon">Bourbon</option>
              <option value="tequila">Tequila</option>
              <option value="gin">Gin</option>
              <option value="rum">Rum</option>
              <option value="whiskey">Whiskey</option>
              <option value="popular">Popular</option>
            </select>
          {favoriteCocktails.length > 0 && <NavLink to='/game-play'>
            <button className="button-quiz-favorites">Quiz me on my favorites</button>
          </NavLink>}
          <p className="Nav_p-favorites">Favorites <br />{favoriteCocktails.length}</p>
        </nav>
    );
  }
}

const mapStateToProps = ({ favoriteCocktails }) => ({
  favoriteCocktails
});

export const mapDispatchToProps = dispatch => ({
  setCurrentCocktails: cocktails => dispatch(setCurrentCocktails(cocktails))
})

export default connect(mapStateToProps, mapDispatchToProps)(Nav);