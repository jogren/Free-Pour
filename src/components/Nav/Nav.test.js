import React from 'react';
import { shallow } from 'enzyme';
import { Nav, mapDispatchToProps } from './Nav';
import { fetchCocktailsBySearch } from '../../apiCalls/apiCalls';
import { setCurrentCocktails } from '../../actions';

jest.mock('../../apiCalls/apiCalls');

describe('Nav', () => {
  let wrapper;
  const mockGetCocktailsByGenre = jest.fn();
  const mockSetCurrentCocktails = jest.fn();
  const mockFavoriteCocktails = [
    {
      name: 'Moscow Mules',
      glass: 'Copper Mug',
      instructions: 'Combine ingredients...',
      image: 'image2.png',
      ingredients: [
        {
          guessed: false,
          ingredient: 'Vodka',
          measure: '2-3 oz'
        },
        {
          guessed: false,
          ingredient: 'Ginger Ale',
          measure: '5oz'
        },
        {
          guessed: false,
          ingredient: 'Lime Juice',
          measure: '2tbsp'
        }
      ]
    }
  ]

  beforeEach(() => {
    wrapper = shallow(<Nav 
      getCocktailsByGenre={mockGetCocktailsByGenre}
      favoriteCocktails={mockFavoriteCocktails}
      setCurrentCocktails={mockSetCurrentCocktails}
    />);
  })

  it('should take a snapshot when passed the correct data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update search state onChange of input', () => {
    wrapper.find('input').at(0).simulate('change', { target: { value: 'margarita', name: 'search' } });
    expect(wrapper.state('search')).toEqual('margarita')
    expect(wrapper.state('searchHasErrored')).toEqual('')
  })

  describe('getCocktailsBySearch', () => {
    it('should invoke fetchCocktailsBySearch', () => {
      wrapper.instance().getCocktailsBySearch();

      expect(fetchCocktailsBySearch).toHaveBeenCalled()
    })

    it('should invoke setCurrentCocktails if fetch is resolved and searchedCocktails is not null', () => {
      fetchCocktailsBySearch.mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve()
        });
      });
      wrapper.instance().getCocktailsBySearch();
      expect(mockSetCurrentCocktails).toHaveBeenCalled()
      expect(wrapper.state('search')).toEqual('')
    });

    it('should reset the search state', () => {
      fetchCocktailsBySearch.mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve()
        });
      });
      wrapper.instance().getCocktailsBySearch();
      expect(wrapper.state('search')).toEqual('')
    });

    it('should return catch error if promise rejects (SAD)', () => {
      fetchCocktailsBySearch.mockImplementation(() => {
        return Promise.reject({
          message: 'Server is down'
        })
      });
      expect(wrapper.instance().getCocktailsBySearch()).rejects.toEqual(Error('Server is down'));
    });

    it('should invoke getCocktailsByGenre onChange of select', () => {
      const mockEvent = { target: { value: 'vodka' }};
      wrapper.find('select').simulate('change', mockEvent);
      expect(mockGetCocktailsByGenre).toHaveBeenCalledWith('vodka');;
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with an setCurrentCocktails action', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentCocktails([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }]);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentCocktails([{ id: 1, name: 'Moscow Mule' }, { id: 2, name: 'Vodka Soda' }]);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
})