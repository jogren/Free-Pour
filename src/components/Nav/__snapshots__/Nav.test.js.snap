// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should take a snapshot when passed the correct data 1`] = `
<nav>
  <div
    className="Nav_div"
  >
    <input
      autoComplete="off"
      className="Nav_input-search"
      name="search"
      onChange={[Function]}
      placeholder="Search for a drink..."
      value=""
    />
    <button
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
  <select
    className="select-container"
    onChange={[Function]}
  >
    <option
      value=""
    >
      Select Genre:
    </option>
    <option
      value="vodka"
    >
      Vodka
    </option>
    <option
      value="bourbon"
    >
      Bourbon
    </option>
    <option
      value="tequila"
    >
      Tequila
    </option>
    <option
      value="gin"
    >
      Gin
    </option>
    <option
      value="rum"
    >
      Rum
    </option>
    <option
      value="whiskey"
    >
      Whiskey
    </option>
    <option
      value="popular"
    >
      Popular
    </option>
  </select>
  <p
    className="Nav_p-favorites"
  >
    Favorites 
    1
  </p>
  <NavLink
    to="/game-play"
  >
    <button
      className="button-quiz-favorites"
    >
      Quiz me on my favorites
    </button>
  </NavLink>
</nav>
`;
